<launch>

    <!-- Static transform from left camera to LIDAR -->
    <node pkg="tf2_ros" type="static_transform_publisher" name="zed_left_camera_optical_frame_to_velodyne"
        args="5.214154202176700392e-02 -1.846812214940576902e-01 -4.561479952008089256e-02 -3.046857448372968236e+00 -1.631619169438448314e+00 -1.657808369943630566e+00 zed_left_camera_optical_frame velodyne"/>

    <!-- Arguments -->
    
    <!-- load_db - if load database -->
    <arg name="load_db" default="false"/>
    <arg if="$(arg load_db)" name="rtabmap_args" value=""/>
    <arg unless="$(arg load_db)" name="rtabmap_args" value="--delete_db_on_start"/>
    
    <!-- db_path - path to database -->
    <arg name="db_path" default="~/.ros/rtabmap.db"/>
    
    <!-- odom - odometry to use -->
    <arg name="odom" default="openvslam"/>
    <group if="$(eval odom == 'openvslam')">
        <!-- Custom node to convert OpenVSLAM odom from camera frame to base_link and publish dynamic transform from odom to base_link -->
        <node pkg="rtabmap_example" type="openvslam_odom_to_base_link.py" name="openvslam_odom_to_base_link" output="screen"/>
    </group>
    <arg if="$(eval odom == 'openvslam')" name="odom_topic" value="/OpenVSLAM/odom_base_link"/>
    <arg if="$(eval odom == 'rtabmap_stereo')" name="odom_topic" value="/rtabmap_stereo_odom"/>
    <arg if="$(eval odom == 'rtabmap_rgbd')" name="odom_topic" value="/rtabmap_rgbd_odom"/>

    <!-- Rtabmap mapping -->
    <node pkg="rtabmap_ros" type="rtabmap" name="rtabmap" output="screen" args="$(arg rtabmap_args)">
        <!-- remap топиков для чтения данных -->
        <remap from="odom" to="$(arg odom_topic)"/>

        <remap from="left/image_rect" to="/zed_node/left/image_rect_color"/>
        <remap from="left/camera_info" to="/zed_node/left/camera_info"/>
        <remap from="right/image_rect" to="/zed_node/right/image_rect_color"/>
        <remap from="right/camera_info" to="/zed_node/right/camera_info"/>
        
        <remap from="scan_cloud" to="/velodyne_points"/>
        
        <remap from="rgb/image" to="/zed_node/left/image_rect_color"/>
        <remap from="rgb/camera_info" to="/zed_node/left/camera_info"/>
        <remap from="depth/image" to="/zed_node/depth/depth_registered"/>
        
        <!-- remap сервисов (для rviz) -->
        <remap from="get_map_data" to="/rtabmap/get_map_data"/>

        <!-- Данные, на которые подписывается rtabmap -->
        <param name="subscribe_stereo" value="false"/>
        <param name="subscribe_depth" value="true"/>
        <param name="subscribe_rgbd" value="false"/>
        <param name="subscribe_scan" value="false"/>
        <param name="subscribe_scan_cloud" value="true"/>
        
        <!-- Путь к database -->
        <param name="database_path" value="$(arg db_path)"/>
        
        <!-- Параметры синхронизации сообщений -->
        <param name="approx_sync" value="true"/>
        <param name="queue_size" value="10"/>
        
        <!-- Параметры rtabmap -->
        <param name="Rtabmap/DetectionRate" value="100"/>
        <param name="Grid/RayTracing" value="true"/>
        <param name="Grid/FromDepth" value="false"/>
    </node>

</launch>

